core/primitives/src/validator_mandates/compute_price.rs

```diff
@@ -0,0 +1,248 @@
+        let config = ValidatorMandatesConfig::new(target_mandates_per_shard, num_shards);
+        let price = compute_mandate_price(config, &stakes);
+        assert_eq!(count_whole_mandates(&stakes, price), target_mandates_per_shard + 11);
+
+        let target_mandates_per_shard = stakes.len() / 2;
+        let config = ValidatorMandatesConfig::new(target_mandates_per_shard, num_shards);
+        let price = compute_mandate_price(config, &stakes);
+        assert_eq!(count_whole_mandates(&stakes, price), target_mandates_per_shard);
+    }
+
+    // Test cases where the stake distribution is exponential.
+    #[test]
+    fn test_exp_dist() {
+        let stakes = {
+            let mut buf = vec![1_000_000_000_u128; 210];
+            let mut last_stake = buf[0];
+            for s in buf.iter_mut().skip(1) {
+                last_stake = last_stake * 97 / 100;
+                *s = last_stake;
+            }
+            buf
+        };
+
+        // This case is similar to the mainnet data.
+        let num_shards = 6;
+        let target_mandates_per_shard = 68;
+        let config = ValidatorMandatesConfig::new(target_mandates_per_shard, num_shards);
+        let price = compute_mandate_price(config, &stakes);
+        assert_eq!(count_whole_mandates(&stakes, price), target_mandates_per_shard * num_shards);
+
+        let num_shards = 1;
+        let target_mandates_per_shard = stakes.len();
+        let config = ValidatorMandatesConfig::new(target_mandates_per_shard, num_shards);
+        let price = compute_mandate_price(config, &stakes);
+        assert_eq!(count_whole_mandates(&stakes, price), target_mandates_per_shard);
+
+        let target_mandates_per_shard = stakes.len() * 2;
+        let config = ValidatorMandatesConfig::new(target_mandates_per_shard, num_shards);
+        let price = compute_mandate_price(config, &stakes);
+        assert_eq!(count_whole_mandates(&stakes, price), target_mandates_per_shard);
+
+        let target_mandates_per_shard = stakes.len() / 2;
+        let config = ValidatorMandatesConfig::new(target_mandates_per_shard, num_shards);
+        let price = compute_mandate_price(config, &stakes);
+        assert_eq!(count_whole_mandates(&stakes, price), target_mandates_per_shard);
+    }
+
+    // Test cases where the stakes are chosen uniformly at random.
+    #[test]
+    fn test_rand_dist() {
+        let stakes = {
+            let mut stakes = vec![0_u128; 1000];
+            let mut rng = rand::rngs::StdRng::seed_from_u64(0xdeadbeef);
+            for s in stakes.iter_mut() {
+                *s = rng.gen_range(1_u128..10_000u128);
+            }
+            stakes
+        };
+
+        let num_shards = 1;
+        let target_mandates_per_shard = stakes.len();
+        let config = ValidatorMandatesConfig::new(target_mandates_per_shard, num_shards);
+        let price = compute_mandate_price(config, &stakes);
+        assert_eq!(count_whole_mandates(&stakes, price), target_mandates_per_shard + 3);
```

Comment:
I couldn't understand how to explain `+ 3` probabilistically, may worth a comment

Reply:
I added a comment in https://github.com/near/nearcore/pull/11044/commits/532663027578c53a4afd29ca4ff49906cfee4e78

It's not very explanatory, but does point out that it is sometimes impossible to get the exact target because we are working with integers.